### DISCUSSION TITLE
PyMOL scripting. How to select the last residue?

### DISCUSSION THREAD

**Discussion by ivandatasci** [schrodinger/pymol-open-source#358] [Q&A] [answered]
Hello PyMOLers.

How to select, say, the last residue of chain C?
Notice that I am asking about the last residue intentionally avoiding selecting it by number.

Thank you.

Ivan

**Comment by chmnk** [comment]
> Notice that I am asking about the last residue intentionally avoiding selecting it by number.

Does it mean that the last residue may have a residue id smaller than the previous residue? That sounds not very conventional.

1) If not, you can find the residue with the largest number (int "resv" instead of the str "resi" may be more convenient in this case). If you have ligands in the same chain with greater residue ids, you may search over "not het" or only over the protein segment if you have a cif file with proper annotations or segment ids specified in your pdb file.
```python
myspace = {"res_ids": set()}
cmd.iterate("chain C and not het", "res_ids.add(resv)", space=myspace)
my_res_id = max(myspace["res_ids"])
cmd.select("my_terminal_res", f"resi {my_res_id}")
```

2) Otherwise, you can use the OXT atom name reserved for the terminal oxygen
```python
myspace = {"res_ids": set()}
cmd.iterate("chain C and name OXT and not het", "res_ids.add((resi, resn, chain))", space=myspace)
my_res_id = myspace["res_ids"].pop()
cmd.select("my_terminal_res", "resi {0} and resn {1} and chain {2}".format(*my_res_id))
```

3) You can manually check the connectivity between N and C atoms residue by residue.

**Comment by speleo3** [comment, marked as answer]
There is a `last` operator.
```
select last_res_in_C, byres (last (chain C))
```
I would also add `guide` and `present` to make sure it's a polymer residue and has coordinates in the current state:
```
select last_res_in_C, byres (last (chain C & present & guide))
```
See https://pymolwiki.org/index.php/Selection_Algebra
